src/lib/homestore_backend/pg_blob_iterator.cpp

```diff
@@ -36,102 +82,178 @@ PG* HSHomeObject::PGBlobIterator::get_pg_metadata() {
+                                LOGE(
+                                    "corrupted blob found, shard_id: {}, blob_id: {}, hash mismatch header [{}] [computed={:np}]",
+                                    shard_id, blob_id, header->to_string(),
+                                    spdlog::to_hex(computed_hash, computed_hash + BlobHeader::blob_max_hash_len));
+                                state = ResyncBlobState::CORRUPTED;
+                            }
+
+                            LOGD("Blob get success: shard_id={}, blob_id={}", shard_id, blob_id);
+                            return std::move(read_buf);
+                        });
 }
 
-void HSHomeObject::PGBlobIterator::create_blobs_snapshot_data(std::vector< BlobInfoData >& blob_data_vec,
-                                                              sisl::io_blob_safe& data_blob, bool end_of_shard) {
-    // std::vector< ::flatbuffers::Offset< BlobData > > blob_entries;
-    // flatbuffers::FlatBufferBuilder builder;
-    // for (auto& b : blob_data_vec) {
-    //     blob_entries.push_back(
-    //         CreateBlobData(builder, b.shard_id, b.blob_id, b.blob.user_key.size(), b.blob.body.size(),
-    //                        builder.CreateVector(r_cast< uint8_t* >(const_cast< char* >(b.blob.user_key.data())),
-    //                                             b.blob.user_key.size()),
-    //                        builder.CreateVector(b.blob.body.bytes(), b.blob.body.size())));
-    // }
-    // builder.FinishSizePrefixed(
-    //     CreateResyncBlobDataBatch(builder, builder.CreateVector(blob_entries), end_of_shard /* end_of_batch */));
-    // data_blob = sisl::io_blob_safe{builder.GetSize()};
-    // std::memcpy(data_blob.bytes(), builder.GetBufferPointer(), builder.GetSize());
+bool HSHomeObject::PGBlobIterator::create_blobs_snapshot_data(sisl::io_blob_safe& data_blob) {
+    std::vector< ::flatbuffers::Offset< ResyncBlobData > > blob_entries;
+    flatbuffers::FlatBufferBuilder builder;
+
+    bool end_of_shard = false;
+    uint64_t total_bytes = 0;
+    auto idx = (uint64_t)(last_end_blob_idx + 1);
+
+    while (total_bytes < max_batch_size && idx < cur_blob_list.size()) {
+        auto info = cur_blob_list[idx];
+        ResyncBlobState state = ResyncBlobState::NORMAL;
+        //handle deleted object
+        if (info.pbas == tombstone_pbas) {
+            state = ResyncBlobState::DELETED;
+            LOGD("Blob is deleted: shard_id={}, blob_id={}, blkid={}", info.shard_id, info.blob_id,
+                 info.pbas.to_string());
+            blob_entries.push_back(CreateResyncBlobDataDirect(builder, info.blob_id, (uint8_t)state, nullptr));
+            continue;
+        }
+        auto result = load_blob_data(info, state).get();
+        if (result.hasError() && result.error().code == BlobErrorCode::READ_FAILED) {
+            LOGW("Failed to retrieve blob for shard={} blob={} pbas={}, err={}", info.shard_id, info.blob_id,
```

Comment:
should we make the state as corrupted? 
should we retry ?

Reply:
Let's make invalid header/mismatched checksum/mismatched shardId as corrupted, and others return BlobErrorCode::READ_FAILED. 
Add retry for BlobErrorCode::READ_FAILED.